README.md
# __DGGRID__

## General Information

__DGGRID__ version 7.0b released September 1, 2019  
Southern Terra Cognita Laboratory / www.discreteglobalgrids.org  


__DGGRID__ is a command-line application designed to generate and manipulate 
icosahedral discrete global grids (DGGs).

Two directories should be included herein:

- `src`: contains complete source code for __DGGRID__ (see `README.MAKE` for 
     instructions on building the application and source code documenation)

- `examples`: contains examples of using __DGGRID__ with pre-computed outputs

User documentation for __DGGRID__ is in `dggridManualV70.pdf`.

## Terms of Use

This documentation is part of __DGGRID__.

__DGGRID__ is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

__DGGRID__ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with the __DGGRID__ source code (see the file `LICENSE`).  If not, see <https://www.gnu.org/licenses/>.

## Credits

__DGGRID__ was primarily written in C++ by Kevin Sahr. See the file `CHANGELOG.md` 
for additional contributors. 

The original __DGGRID__ specifications were developed by (in alphabetical order): 
A. Ross Kiester, Tony Olsen, Barbara Rosenbaum, Kevin Sahr, Ann Whelan, and 
Denis White.

__DGGRID__ was made possible in part by funding from the __US Environmental Protection Agency__ and __Culmen International__.

## Dependencies

__DGGRID__ requires the following external library (not included):

- The Open Source Geospatial Foundation’s GDAL translator library for raster and vector geospatial data formats (see http://www.gdal.org)

__DGGRID__ uses the following external libraries (included with the __DGGRID__ source 
code):

- Angus Johnson’s Clipper library; see http://www.angusj.com.

- George Marsaglia’s multiply-with-carry “Mother-of-all-RNGs” pseudo-random number generation function.





CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

All changes are by Kevin Sahr, unless otherwise noted.

## [7.0] - 2019-08-01
### Added
- added CULMEN/PLANETRISK choice to parameter dggs_type 
- added ISEA7H and FULLER7H choices to parameter dggs_type 
- added aperture 7 choice to parameter dggs_aperture 
- added SEQUENCE choice to parameter dggs_aperture_type
- added new parameter dggs_aperture_sequence
- added GDAL choice to parameter clip_subset_type
- added GDAL choice to parameters cell_output_type and point_output_type
- added new parameter cell_output_gdal_format
- added new parameter point_output_gdal_format
- added cell neighbor output, including parameters neighbor_output_type and neighbor_output_file_name
- added cell children output, including parameters children_output_type and children_output_file_name
- added parameters output_first_seqnum and output_last_seqnum
- restored INTERLEAVE choice to parameter output_address_type with corrected error messages
### Changed
- released DGGRID under the AGPL license
- gnomonic projection now calculated manually, allowing the removal of the PROJ4 library dependency
- made DgSerialConverter conversion trace code more informative/legible
- updated documentation and examples to reflect all changes

## [6.51] - 2019-07-01
### Changed
- simplified the point-on-icosahedron-face function

## [6.5] - 2019-05-31
### Changed
- changed the rotation angle from backing frame to substrate grid backing frame for a DgHexC2Grid2D from 30.0 to -30.0, to correspond to the definition in the PlanetRisk code fork. Note that cell boundary polygons are generated beginning with a different vertex than in previous versions.
- changed examples/sampleOutput to reflect the rotation angle change
- minor changes to examples output formatting

## [6.42] - 2019-05-03
### Changed
- using shortest distance to face center point to calculate the face a point is on
- all trig functions and sqrt's now use long double versions
- precalculated more constants

## [6.41] - 2019-04-19
### Fixed
- error in coslon limiting in DgEllipsoidRF.cpp (thanks to @billyauhk who
found the error in H3)

## [6.4] - 2018-09-18
### Added
- added parameter clipper_scale_factor
- added SEQNUMS choice to parameter clip_subset_type (thanks to R. Barnes)
- example seqnums
### Fixed
- shapelib header file directory include in MakeIncludes
- overage in I and J intersection bug
### Changed
- added bounding box coarse filter before clipper polygon intersection

## [6.3] - 2018-09-01
### Added
- this file CHANGELOG.md
### Fixed
- gcc v6 compiler errors (from R. Barnes)
- llvm compile error in DgLocVector.h
### Changed
- released under an AGPL license
- using clipper for polygon intersection (thanks to R. Barnes)
- cleaned-up examples
### Deleted
- unused hypot.c from proj4lib

## [6.2b] - 2018-08-31
### Added
- Created private github repo.

## [6.2] - 2015-08-30
- restored v4.3b TRANSFORM_POINTS operation and supporting parameters
- restored v4.3b BIN_POINT_VALS operation and supporting parameters
- restored v4.3b BIN_POINT_PRESENCE operation and supporting parameters
- added GeoJSON output format (Matt Gregory)
- added parameters kml_name and kml_description
- added new examples
- corrected/suppressed (as applicable) various compiler warnings
- added the metafile editor program MFeditor.jar (available
  separately from www.discreteglobalgrids.org); MFeditor was
  written by Michael Paradis and Benjamin Harris, and updated 
  for DGGRID 6.2b by Jeremy Anders and Anthony Serna.

## [6.1] - 2013-05-28
- fixed vertex ordering in ArcInfo Generate file output 
  polygons (bug introduced in version 6.02b)
- fixed bug that caused Superfund cells that straddle
  quad boundaries to not be appropriately clipped in some
  situations
- added parameters kml_default_color and kml_default_width

## [6.02b] - 2013-04-07
- corrected reversed winding order on output shapefile polygons
- added dglib code usage demo in src/apps/appex
- completed manual appendices A (DGGRID Metafile Parameters) and C 
  (Characteristics of DGGRID Hexagonal DGGs)
- changed output_cell_label_type parameter value "LINEAR_INDEX"
  to the more descriptive value "GLOBAL_SEQUENCE"
- for naming consistency changed parameter concatenate_randpts_output 
  to randpts_concatenate_output and num_randpts_per_cell to 
  randpts_num_per_cell
- changed DgParamList.h to correctly handle strict c++ two-pass
  template compiling without -fpermissive 
- fixed rare crashes caused by abstract classes with missing
  virtual destructors
- numerous small changes to allow or improve compilation on a
  variety of platforms

## [6.0b] - 2013-01-29

- increased precision throughout allowing for higher
  resolution DGGs
- added preset DGGS parameter types
- added mixed aperture 4 and 3 hexagon DGGSs, greatly 
  increasing the number of grid choices
- can generate DGGSs using Fuller's icosahedral projection (projection 
  was written in R by Denis White and then ported to C++ by James 
  Scharmann)
- includes the Superfund_500m hierarchically indexed mixed
  aperture DGGS
- added ESRI Shapefile format for input and output
- added KML format for output (Jesse Williamson)
- added ability to break-up large output files by setting
  the max number of cells to write to a single file
- added operation that outputs DGGS characteristics table
- improved feedback during execution
- improved Makefiles
- builds cleanly on MacOS 10.8.2 using gcc 4.2.1 and on 
  Ubuntu 11.04 with gcc 4.5.2 (Jesse Williamson)

## [4.3b] - 2003-06-21
- first public release

## [3.1b] - 2001-11-11

## [0.9b] - 2001-10-01
- initial port of Kevin Sahr's tclib to dglib
- Lian Song wrote many of the spherical trigonometry routines and the 
  original implementation of the ISEA projection.


src/apps/dggrid/dggrid.h
/*******************************************************************************
    Copyright (C) 2018 Kevin Sahr

    This file is part of DGGRID.

    DGGRID is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    DGGRID is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*******************************************************************************/
////////////////////////////////////////////////////////////////////////////////
//
// dggrid.h: this file defines the classes (effectively structures) which hold
//           the parameter info
//
// Version 6.1 - Kevin Sahr, 5/23/13
// Version 6.2 - Kevin Sahr, 10/25/14
// Version 7.0 - Kevin Sahr, 9/1/19
//
////////////////////////////////////////////////////////////////////////////////

#ifndef DGGRID_H
#define DGGRID_H

#include <iostream>
#include <set>
#include "DgParamList.h"
#include "DgEllipsoidRF.h"
#include "DgRandom.h"
#include "DgOutLocFile.h"
#include "DgIDGG.h"
#include "DgInShapefileAtt.h"
#include "DgApSeq.h"
#include "util.h"
#include "DgRunningStats.h"

using namespace std;

class DggridParam;
class DgEvalData;
class DgOutShapefile;
class DgOutPRCellsFile;
class DgOutPRPtsFile;
class DgOutNeighborsFile;
class DgOutChildrenFile;
class DgOutputStream;

// constants
#define MAX_DGG_RES 35

////////////////////////////////////////////////////////////////////////////////
//// function prototypes ////

// gridgen.cpp

class MainParam;
class DgGridPList;
void orientGrid (MainParam& dp, DgGridPList& plist);

// gridgen.cpp

class GridGenParam;
class DgGridPList;
void outputStatus (const GridGenParam& dp, bool force = false);
void doGridGen (GridGenParam& dp, DgGridPList& plist);

// binvals.cpp

class BinValsParam;
void doBinVals (BinValsParam& dp, DgGridPList& plist);

// binpres.cpp

class BinPresenceParam;
void doBinPresence (BinPresenceParam& dp, DgGridPList& plist);

// transform.cpp

class TransformParam;
void doTransforms (TransformParam& dp, DgGridPList& plist);

// table.cpp

class MainParam;
void doTable (MainParam& dp, DgGridPList& plist);

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
//// the parameter structures ////
// no attempt to be subtle here; these contain the "global" parameters for
// the specified dggrid operation 

// common parameters

class MainParam {

   public:

      MainParam (DgParamList& plist);

      virtual ~MainParam();

      void dump (void);   // output to cout

      // the parameters

      string operation;             // operation for dggrid to perform
      string dggsType;              // preset DGGS type
      string gridTopo;              // diamond/hex/triangle
      int aperture;      // aperture
      string projType;              // projection type
      int res;           // resolution (may be adjusted)
      int actualRes;     // original, actual resolution
      bool placeRandom;             // random grid placement?
      bool orientCenter;            // grid placement based on quad center point?
      DgRandom* orientRand;             // RNG for generating random orientation
      int numGrids;      // number of grids to generate
      int curGrid;       // grid counter
      bool lastGrid;                    // last grid?
      DgGeoCoord vert0;                 // placement vert 
      long double azimuthDegs;               // orientation azimuth
      long double earthRadius;               // earth radius in km
      string datum;               // datum used to determine the earthRadius
      int precision;     // number of digits after decimal pt to output
      int verbosity;     // debugging info verbosity
      bool megaVerbose;                 // 
      bool useMother;                   // use Mother RNG?
      string metaOutFileNameBase;
      string metaOutFileName;
      string apertureType;		// "PURE", "MIXED43", "SUPERFUND", or "SEQUENCE"
      bool   isMixed43;		// are we using mixed43 aperture?
      int numAp4;      // # of leading ap 4 resolutions in a mixed grid
      bool   isSuperfund;
      bool   isApSeq;		// are we using an aperture sequence?
      DgApSeq apSeq;
      int   sfRes; // superfund digit resolution

   protected:

      void determineRes (const DgParamList& plist);
      DgGridStats genStats (int res);

};

////////////////////////////////////////////////////////////////////////////////
// grid generation parameter structure

class GridGenParam : public MainParam {
   
   public:

      GridGenParam (DgParamList& plist);
      ~GridGenParam ();

      void dump (void);   // output to cout

      // the parameters

      bool wholeEarth;       // generate entire grid?
      bool seqToPoly;        // whether user wants polys from seqnum
      bool useGDAL;          // use GDAL for either input or output
      bool clipAIGen;        // clip using AIGen files (or Shapefiles)
      bool clipGDAL;         // clip using GDAL files
      bool clipShape;        // clip using Shapefiles
      vector<string> regionFiles;
      int nRandPts;          // # of random pts generated for each hex
      bool clipRandPts;      // clip randpts to polys
      int nDensify;          // number of points-per-edge of densification
      long double nudge;        // adjustment for quad intersection consistency
      DgRandom* ptsRand;     // RNG for generating random points

      string cellOutType;
      string gdalCellDriver;
      string pointOutType;
      string gdalPointDriver;
      string randPtsOutType;

      string neighborsOutType;
      string childrenOutType;
      string neighborsOutFileNameBase;
      string neighborsOutFileName;
      string childrenOutFileNameBase;
      string childrenOutFileName;

      string cellOutFileNameBase;
      string cellOutFileName;
      string ptOutFileNameBase;
      string ptOutFileName;
      string randPtsOutFileNameBase;
      string randPtsOutFileName;
      int    shapefileIdLen;  // global_id string field length

      string kmlColor;
      int    kmlWidth;
      string kmlName;
      string kmlDescription;

      bool doPointInPoly;             // perform pt-in-poly intersection
      bool doPolyIntersect;           // perform poly intersection
      long long int sampleCount; // last sample point sequence number
      unsigned long long int nSamplePts;
      bool doRandPts;                 // generate random points for the cells
      unsigned long int clipperFactor; // clipper scaling factor
      long double invClipperFactor; // 1.0L / clipper scaling factor

      DgOutLocFile *cellOut, *ptOut, *randPtsOut;
      DgOutShapefile *cellOutShp, *ptOutShp;
      DgOutPRCellsFile *prCellOut;
      DgOutPRPtsFile *prPtOut;
      DgOutNeighborsFile *nbrOut;
      DgOutChildrenFile *chdOut;

      DgRunningStats runStats;

      bool concatPtOut;         
      char formatStr[50];         
      bool useEnumLbl;         
      unsigned long long int nCellsTested; 
      unsigned long long int nCellsAccepted;         
      unsigned long long int nCellsOutputToFile; // cells output to current file 
      unsigned long int nOutputFile; // # of current output file

      unsigned long int updateFreq; // how often to output updates         
      unsigned long int maxCellsPerFile; // max cells in a single output file
      unsigned long int outFileStartNum; // multi-part file number to begin generation
      unsigned long int outFileLastNum; // multi-part file number where generation ends

      long double geoDens;          // max arc length in radians

      bool buildShapeFileAttributes; // create fields for shapefile output
      bool buildClipFileAttributes;  // use clipping shape files (vs. others)
     
      int shapefileDefaultInt;
      long double shapefileDefaultDouble;
      string shapefileDefaultString;

      vector<string> attributeFiles;
      bool outCellAttributes; // true if cell output file is a shapefile
      bool outPointAttributes; // true if point output file is a shapefile
      set<DgDBFfield> allFields; // union of all input file attribute fields
      set<DgDBFfield> curFields; // union of all intersected fields for the
                                 // current cell
};

////////////////////////////////////////////////////////////////////////////////
// bin point values parameter structure

class BinValsParam : public MainParam {
   
   public:

      BinValsParam (DgParamList& plist);
     
      ~BinValsParam ();

      void dump (void);   // output to cout

      // the parameters

      bool wholeEarth;
      string outFileName;
      string outFileNameBase;
      ofstream* outFile;
      vector<string> inputFiles;
      string outAddType;
      bool outSeqNum;
      char inputDelimiter;
      char outputDelimiter;
      string inFormatStr;
      bool outputAllCells;   // or only occupied ones?
};

////////////////////////////////////////////////////////////////////////////////
// bin point presence parameter structure

class BinPresenceParam : public MainParam {
   
   public:

      BinPresenceParam (DgParamList& plist);

      ~BinPresenceParam ();

      void dump (void);   // output to cout

      // the parameters

      bool wholeEarth;
      string outFileName;
      string outFileNameBase;
      ofstream* outFile;
      vector<string> inputFiles;
      string outAddType;
      bool outSeqNum;
      char inputDelimiter;
      char outputDelimiter;
      string inFormatStr;
      bool outputAllCells;   // or only occupied ones?
      bool outputCount;

};

////////////////////////////////////////////////////////////////////////////////
// transform addresses parameter structure

class TransformParam : public MainParam {
   
   public:

      TransformParam (DgParamList& plist);

      void dump (void);   // output to cout

      // the parameters

      string inFileName;
      string inAddType;

      bool inSeqNum;
      bool outSeqNum;
      char inputDelimiter;
      char outputDelimiter;

      string outFileName;
      string outFileNameBase;
      string outFileNameAIG;
      string outFileNameKML;
      string outAddType;

      int nDensify;

};

////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
class DgGridPList : public DgParamList {

   public:

      DgGridPList (void);

   private:

      void init2 (void);

};

////////////////////////////////////////////////////////////////////////////////

#endif
